pytorch_lightning/trainer/training_loop.py

```diff
@@ -676,26 +678,33 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
                     model = self.trainer.get_model()
                     model.toggle_optimizer(optimizer, opt_idx)
 
-                if not (accumulation_done or is_final_batch):
+                if should_accumulate:
                     # For gradient accumulation
 
                     # -------------------
                     # calculate loss (train step + train step end)
                     # -------------------
-                    self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
+                    # perform dpp sync only when performing optimizer_step
+                    with self.block_ddp_sync_behaviour():
+                        self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
```

Comment:
do we need `block_ddp_sync_behaviour` ? 
```suggestion
                    # perform dpp sync only when performing optimizer_step
                    if isinstance(self.trainer.model, torch.nn.parallel.DistributedDataParallel):
                        with self.trainer.model.no_sync():
                            self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
                     else:
                          self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
                    with self.block_ddp_sync_behaviour():
                        self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
```

Reply:
Just to make it more explicit for our readers :)

Reply:
@tchaton requested this to make it more readable and hide the conditions for that in the context manager

Reply:
@ananthsub 
The training and evaluation loop should look absolutely perfect and simple to understand.
As a new coder, I should recognise my training loop. As a new coder, I have no knowledge about ddp and your suggested change would have confused me :)

One great example of this PR is: not (accumulation_done or is_final_batch) -> should_accumulate. It is pretty simple, but make a clear statement about what is happening. We should try to enforce this as much as possible :)

I hope it makes sense :)